### ISSUE TITLE
Cached capabilities expire too soon?

### ISSUE THREAD

**Issue by SystemKeeper** [strukturag/nextcloud-spreed-signaling#782]
* Related https://github.com/strukturag/nextcloud-spreed-signaling/pull/780

`curl -ik -H "OCS-ApiRequest: true" -H "Accept: application/json" https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities` shows:

```
HTTP/2 200 
server: nginx
date: Thu, 25 Jul 2024 18:25:45 GMT
content-type: application/json; charset=utf-8
content-length: 3972
referrer-policy: no-referrer
x-content-type-options: nosniff
x-frame-options: SAMEORIGIN
x-permitted-cross-domain-policies: none
x-robots-tag: noindex, nofollow
x-xss-protection: 1; mode=block
x-powered-by: PHP/8.1.27
...
x-debug-token: NfvOqsLLzNFasGjBUjUH
server-timing: token;desc="NfvOqsLLzNFasGjBUjUH"
x-request-id: NfvOqsLLzNFasGjBUjUH
cache-control: no-cache, no-store, must-revalidate
feature-policy: autoplay 'none';camera 'none';fullscreen 'none';geolocation 'none';microphone 'none';payment 'none'
etag: "656c8fa3fd1c85f2590d63c92a6328d9"
```

So currently the `cache-control` header is set to `no-cache`, which seems exactly what is happening here: 

```
2024-07-25 19:28:13 capabilities.go:288: Capabilities expired for https://nextcloud2-mm.local/ocs/v2.php/cloud/capabilities, updating
2024-07-25 19:28:13 capabilities.go:125: Capabilities map[config:map[attachments:map[allowed:false] call:map[breakout-rooms:true can-enable-sip:false can-upload-background:false enabled:true predefined-backgrounds:[1_office.jpg 2_home.jpg 3_abstract.jpg 4_beach.jpg 5_park.jpg 6_theater.jpg 7_library.jpg 8_space_station.jpg] recording:false recording-consent:0 sip-dialout-enabled:false sip-enabled:false supported-reactions:[❤️ 🎉 👏 👍 👎 😂 🤩 🤔 😲 😥]] chat:map[has-translation-providers:false max-length:32000 read-privacy:0 typing-privacy:0] conversations:map[can-create:false] federation:map[enabled:false incoming-enabled:false only-trusted-servers:true outgoing-enabled:false] previews:map[max-gif-size:3.145728e+06] signaling:map[hello-v2-token-key:-----BEGIN PUBLIC KEY-----
2024-07-25 19:28:13 MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEGSND9JoFDbD2tyMoAfkJ3jQazY9p
2024-07-25 19:28:13 wHZN7o7X1rVB4GfKuWT2qHbCGDU69MRDm/oV92MKJcFCbSCQwPecm/uUAw==
2024-07-25 19:28:13 -----END PUBLIC KEY-----
2024-07-25 19:28:13  session-ping-limit:200]] config-local:map[attachments:[allowed folder] call:[predefined-backgrounds can-upload-background] chat:[read-privacy has-translation-providers typing-privacy] conversations:[can-create] federation:[enabled incoming-enabled outgoing-enabled only-trusted-servers] previews:[max-gif-size] signaling:[session-ping-limit hello-v2-token-key]] features:[audio video chat-v2 conversation-v4 guest-signaling empty-group-room guest-display-names multi-room-users favorites last-room-activity no-ping system-messages delete-messages mention-flag in-call-flags conversation-call-flags notification-levels invite-groups-and-mails locked-one-to-one-rooms read-only-rooms listable-rooms chat-read-marker chat-unread webinary-lobby start-call-flag chat-replies circles-support force-mute sip-support sip-support-nopin chat-read-status phonebook-search raise-hand room-description rich-object-sharing temp-user-avatar-api geo-location-sharing voice-message-sharing signaling-v3 publishing-permissions clear-history direct-mention-flag notification-calls conversation-permissions rich-object-list-media rich-object-delete unified-search chat-permission silent-send silent-call send-call-notification talk-polls breakout-rooms-v1 recording-v1 avatar chat-get-context single-conversation-status chat-keep-notifications typing-privacy remind-me-later bots-v1 markdown-messages media-caption session-state note-to-self recording-consent sip-support-dialout delete-messages-unlimited edit-messages silent-send-state chat-read-last federation-v1 federation-v2 ban-v1 chat-reference-id mention-permissions message-expiration reactions] features-local:[favorites chat-read-status listable-rooms phonebook-search temp-user-avatar-api unified-search avatar remind-me-later note-to-self] version:20.0.0-dev.8] from https://nextcloud2-mm.local/ocs/v2.php/cloud/capabilities have not changed
2024-07-25 19:28:13 capabilities.go:125: Capabilities map[config:map[attachments:map[allowed:false] call:map[breakout-rooms:true can-enable-sip:false can-upload-background:false enabled:true predefined-backgrounds:[1_office.jpg 2_home.jpg 3_abstract.jpg 4_beach.jpg 5_park.jpg 6_theater.jpg 7_library.jpg 8_space_station.jpg] recording:false recording-consent:0 sip-dialout-enabled:false sip-enabled:false supported-reactions:[❤️ 🎉 👏 👍 👎 😂 🤩 🤔 😲 😥]] chat:map[has-translation-providers:false max-length:32000 read-privacy:0 typing-privacy:0] conversations:map[can-create:false] federation:map[enabled:false incoming-enabled:false only-trusted-servers:true outgoing-enabled:false] previews:map[max-gif-size:3.145728e+06] signaling:map[hello-v2-token-key:-----BEGIN PUBLIC KEY-----
2024-07-25 19:28:13 MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAECQ96obd2/VE1fIPZ8ZjZCW3Df4HK
2024-07-25 19:28:13 9kCsFR4wqlLLQ4V2oGLUomTGrpo0ZEReR7z7xlTL0rLPq+kj7f4iYePsJA==
2024-07-25 19:28:13 -----END PUBLIC KEY-----
2024-07-25 19:28:13  session-ping-limit:200]] config-local:map[attachments:[allowed folder] call:[predefined-backgrounds can-upload-background] chat:[read-privacy has-translation-providers typing-privacy] conversations:[can-create] federation:[enabled incoming-enabled outgoing-enabled only-trusted-servers] previews:[max-gif-size] signaling:[session-ping-limit hello-v2-token-key]] features:[audio video chat-v2 conversation-v4 guest-signaling empty-group-room guest-display-names multi-room-users favorites last-room-activity no-ping system-messages delete-messages mention-flag in-call-flags conversation-call-flags notification-levels invite-groups-and-mails locked-one-to-one-rooms read-only-rooms listable-rooms chat-read-marker chat-unread webinary-lobby start-call-flag chat-replies circles-support force-mute sip-support sip-support-nopin chat-read-status phonebook-search raise-hand room-description rich-object-sharing temp-user-avatar-api geo-location-sharing voice-message-sharing signaling-v3 publishing-permissions clear-history direct-mention-flag notification-calls conversation-permissions rich-object-list-media rich-object-delete unified-search chat-permission silent-send silent-call send-call-notification talk-polls breakout-rooms-v1 recording-v1 avatar chat-get-context single-conversation-status chat-keep-notifications typing-privacy remind-me-later bots-v1 markdown-messages media-caption session-state note-to-self recording-consent sip-support-dialout delete-messages-unlimited edit-messages silent-send-state chat-read-last federation-v1 federation-v2 ban-v1 chat-reference-id mention-permissions message-expiration reactions] features-local:[favorites chat-read-status listable-rooms phonebook-search temp-user-avatar-api unified-search avatar remind-me-later note-to-self] version:20.0.0-dev.8] from https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities have not changed
2024-07-25 19:28:15 capabilities.go:288: Capabilities expired for https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities, updating
2024-07-25 19:28:15 capabilities.go:125: Capabilities map[config:map[attachments:map[allowed:false] call:map[breakout-rooms:true can-enable-sip:false can-upload-background:false enabled:true predefined-backgrounds:[1_office.jpg 2_home.jpg 3_abstract.jpg 4_beach.jpg 5_park.jpg 6_theater.jpg 7_library.jpg 8_space_station.jpg] recording:false recording-consent:0 sip-dialout-enabled:false sip-enabled:false supported-reactions:[❤️ 🎉 👏 👍 👎 😂 🤩 🤔 😲 😥]] chat:map[has-translation-providers:false max-length:32000 read-privacy:0 typing-privacy:0] conversations:map[can-create:false] federation:map[enabled:false incoming-enabled:false only-trusted-servers:true outgoing-enabled:false] previews:map[max-gif-size:3.145728e+06] signaling:map[hello-v2-token-key:-----BEGIN PUBLIC KEY-----
2024-07-25 19:28:15 MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAECQ96obd2/VE1fIPZ8ZjZCW3Df4HK
2024-07-25 19:28:15 9kCsFR4wqlLLQ4V2oGLUomTGrpo0ZEReR7z7xlTL0rLPq+kj7f4iYePsJA==
2024-07-25 19:28:15 -----END PUBLIC KEY-----
2024-07-25 19:28:15  session-ping-limit:200]] config-local:map[attachments:[allowed folder] call:[predefined-backgrounds can-upload-background] chat:[read-privacy has-translation-providers typing-privacy] conversations:[can-create] federation:[enabled incoming-enabled outgoing-enabled only-trusted-servers] previews:[max-gif-size] signaling:[session-ping-limit hello-v2-token-key]] features:[audio video chat-v2 conversation-v4 guest-signaling empty-group-room guest-display-names multi-room-users favorites last-room-activity no-ping system-messages delete-messages mention-flag in-call-flags conversation-call-flags notification-levels invite-groups-and-mails locked-one-to-one-rooms read-only-rooms listable-rooms chat-read-marker chat-unread webinary-lobby start-call-flag chat-replies circles-support force-mute sip-support sip-support-nopin chat-read-status phonebook-search raise-hand room-description rich-object-sharing temp-user-avatar-api geo-location-sharing voice-message-sharing signaling-v3 publishing-permissions clear-history direct-mention-flag notification-calls conversation-permissions rich-object-list-media rich-object-delete unified-search chat-permission silent-send silent-call send-call-notification talk-polls breakout-rooms-v1 recording-v1 avatar chat-get-context single-conversation-status chat-keep-notifications typing-privacy remind-me-later bots-v1 markdown-messages media-caption session-state note-to-self recording-consent sip-support-dialout delete-messages-unlimited edit-messages silent-send-state chat-read-last federation-v1 federation-v2 ban-v1 chat-reference-id mention-permissions message-expiration reactions] features-local:[favorites chat-read-status listable-rooms phonebook-search temp-user-avatar-api unified-search avatar remind-me-later note-to-self] version:20.0.0-dev.8] from https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities have not changed
2024-07-25 19:28:15 capabilities.go:288: Capabilities expired for https://nextcloud2-mm.local/ocs/v2.php/cloud/capabilities, updating
2024-07-25 19:28:15 capabilities.go:125: Capabilities map[config:map[attachments:map[allowed:false] call:map[breakout-rooms:true can-enable-sip:false can-upload-background:false enabled:true predefined-backgrounds:[1_office.jpg 2_home.jpg 3_abstract.jpg 4_beach.jpg 5_park.jpg 6_theater.jpg 7_library.jpg 8_space_station.jpg] recording:false recording-consent:0 sip-dialout-enabled:false sip-enabled:false supported-reactions:[❤️ 🎉 👏 👍 👎 😂 🤩 🤔 😲 😥]] chat:map[has-translation-providers:false max-length:32000 read-privacy:0 typing-privacy:0] conversations:map[can-create:false] federation:map[enabled:false incoming-enabled:false only-trusted-servers:true outgoing-enabled:false] previews:map[max-gif-size:3.145728e+06] signaling:map[hello-v2-token-key:-----BEGIN PUBLIC KEY-----
2024-07-25 19:28:15 MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEGSND9JoFDbD2tyMoAfkJ3jQazY9p
2024-07-25 19:28:15 wHZN7o7X1rVB4GfKuWT2qHbCGDU69MRDm/oV92MKJcFCbSCQwPecm/uUAw==
2024-07-25 19:28:15 -----END PUBLIC KEY-----
2024-07-25 19:28:15  session-ping-limit:200]] config-local:map[attachments:[allowed folder] call:[predefined-backgrounds can-upload-background] chat:[read-privacy has-translation-providers typing-privacy] conversations:[can-create] federation:[enabled incoming-enabled outgoing-enabled only-trusted-servers] previews:[max-gif-size] signaling:[session-ping-limit hello-v2-token-key]] features:[audio video chat-v2 conversation-v4 guest-signaling empty-group-room guest-display-names multi-room-users favorites last-room-activity no-ping system-messages delete-messages mention-flag in-call-flags conversation-call-flags notification-levels invite-groups-and-mails locked-one-to-one-rooms read-only-rooms listable-rooms chat-read-marker chat-unread webinary-lobby start-call-flag chat-replies circles-support force-mute sip-support sip-support-nopin chat-read-status phonebook-search raise-hand room-description rich-object-sharing temp-user-avatar-api geo-location-sharing voice-message-sharing signaling-v3 publishing-permissions clear-history direct-mention-flag notification-calls conversation-permissions rich-object-list-media rich-object-delete unified-search chat-permission silent-send silent-call send-call-notification talk-polls breakout-rooms-v1 recording-v1 avatar chat-get-context single-conversation-status chat-keep-notifications typing-privacy remind-me-later bots-v1 markdown-messages media-caption session-state note-to-self recording-consent sip-support-dialout delete-messages-unlimited edit-messages silent-send-state chat-read-last federation-v1 federation-v2 ban-v1 chat-reference-id mention-permissions message-expiration reactions] features-local:[favorites chat-read-status listable-rooms phonebook-search temp-user-avatar-api unified-search avatar remind-me-later note-to-self] version:20.0.0-dev.8] from https://nextcloud2-mm.local/ocs/v2.php/cloud/capabilities have not changed
2024-07-25 19:28:22 client.go:353: Client 4XW9g09SsUOgfo-vWddARFlRx0RiTch2JKHAtBlpAdp8blN0bjJ1VGl6bER4NHFQem8ta3Z4cFZiWXFmQS1OVk55THJKMHQ0T1RWRlpYZTVYTndpUXFnM2RicGY2bkphaDFBTTBZbjJob3RuY085Vk1KbUxUeXJxbGpkVGEyOVdaSmx2UDhsakVXVWtrOU16WmdUeHhyT0xvdjJmOGRMb1RveWFqOXFXVUxxR21wRUF5eVhvdzBmR0pJS1JFeC1QazkweDVDUlllfDQ5MzgyOTEyNzE= has RTT of 0 ms (738.334µs)
2024-07-25 19:28:22 capabilities.go:288: Capabilities expired for https://nextcloud2-mm.local/ocs/v2.php/cloud/capabilities, updating
2024-07-25 19:28:22 capabilities.go:288: Capabilities expired for https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities, updating
2024-07-25 19:28:23 capabilities.go:125: Capabilities map[config:map[attachments:map[allowed:false] call:map[breakout-rooms:true can-enable-sip:false can-upload-background:false enabled:true predefined-backgrounds:[1_office.jpg 2_home.jpg 3_abstract.jpg 4_beach.jpg 5_park.jpg 6_theater.jpg 7_library.jpg 8_space_station.jpg] recording:false recording-consent:0 sip-dialout-enabled:false sip-enabled:false supported-reactions:[❤️ 🎉 👏 👍 👎 😂 🤩 🤔 😲 😥]] chat:map[has-translation-providers:false max-length:32000 read-privacy:0 typing-privacy:0] conversations:map[can-create:false] federation:map[enabled:false incoming-enabled:false only-trusted-servers:true outgoing-enabled:false] previews:map[max-gif-size:3.145728e+06] signaling:map[hello-v2-token-key:-----BEGIN PUBLIC KEY-----
2024-07-25 19:28:23 MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAECQ96obd2/VE1fIPZ8ZjZCW3Df4HK
2024-07-25 19:28:23 9kCsFR4wqlLLQ4V2oGLUomTGrpo0ZEReR7z7xlTL0rLPq+kj7f4iYePsJA==
2024-07-25 19:28:23 -----END PUBLIC KEY-----
2024-07-25 19:28:23  session-ping-limit:200]] config-local:map[attachments:[allowed folder] call:[predefined-backgrounds can-upload-background] chat:[read-privacy has-translation-providers typing-privacy] conversations:[can-create] federation:[enabled incoming-enabled outgoing-enabled only-trusted-servers] previews:[max-gif-size] signaling:[session-ping-limit hello-v2-token-key]] features:[audio video chat-v2 conversation-v4 guest-signaling empty-group-room guest-display-names multi-room-users favorites last-room-activity no-ping system-messages delete-messages mention-flag in-call-flags conversation-call-flags notification-levels invite-groups-and-mails locked-one-to-one-rooms read-only-rooms listable-rooms chat-read-marker chat-unread webinary-lobby start-call-flag chat-replies circles-support force-mute sip-support sip-support-nopin chat-read-status phonebook-search raise-hand room-description rich-object-sharing temp-user-avatar-api geo-location-sharing voice-message-sharing signaling-v3 publishing-permissions clear-history direct-mention-flag notification-calls conversation-permissions rich-object-list-media rich-object-delete unified-search chat-permission silent-send silent-call send-call-notification talk-polls breakout-rooms-v1 recording-v1 avatar chat-get-context single-conversation-status chat-keep-notifications typing-privacy remind-me-later bots-v1 markdown-messages media-caption session-state note-to-self recording-consent sip-support-dialout delete-messages-unlimited edit-messages silent-send-state chat-read-last federation-v1 federation-v2 ban-v1 chat-reference-id mention-permissions message-expiration reactions] features-local:[favorites chat-read-status listable-rooms phonebook-search temp-user-avatar-api unified-search avatar remind-me-later note-to-self] version:20.0.0-dev.8] from https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities have not changed
2024-07-25 19:28:23 capabilities.go:288: Capabilities expired for https://nextcloud-mm.local/ocs/v2.php/cloud/capabilities, updating
...
```

Just wondering, that is what we expect here, right? I mean in this case it's just the development setup, but since we don't set it explicitly in code (https://github.com/strukturag/nextcloud-spreed-signaling/pull/780#issuecomment-2248129713) we have to rely on some fallback on the server side which might not be there.

**Comment by fancycode** [comment]
The signaling server (almost) does what the `Cache-Control` header says:
- `no-cache` states that the response may be stored in the cache but must be validated with the origin server before it is used.
- `no-store` states the response should not be stored (this is ignored).
- `must-revalidate` states that if the validation of a cached response fails, the stale cache entry may not be used.

So it's expected that the capabilities are updated on each access. However as the `ETag` header is also returned, the signaling server will validate the cached entry with `If-None-Match` which will return `304 Not Modified` and omit the actual capabilities data.

Thinking about it, we probably still should cache for at least one second or so, as the capabilities are used a couple of times when a client connects (check some feature flags, get the token validation key, etc.) which in this case will validate the data on each access.

**Comment by SystemKeeper** [comment]
Yea, this seems to be the default value for responses: 
https://github.com/nextcloud/server/blob/97ea95714aeec2d23567ae5d85b28fd6b688b1cc/lib/public/AppFramework/Http/Response.php#L235

On c.nc.c it is the same. Not sure where you saw the `max-age=60` in your tests?
I think we really make sure that we have a sensible cache-control header on the nextcloud instance for that as it feels like this can be problematic, when all instance by default return `no-cache`.

**Comment by fancycode** [comment]
> Not sure where you saw the `max-age=60` in your tests?

It was an old recommendation to set a default expiration of 60 seconds (see https://github.com/strukturag/nextcloud-spreed-signaling/pull/780#issuecomment-2248167289) and I still had this in my dev environment configured.

> I think we should really make sure that we have a sensible cache-control header on the nextcloud instance for that, as it feels like this can be problematic when all instance by default return `no-cache`.

Yeah, but this is out of scope for this issue here ;-) The signaling server should comply with whatever is returned by Nextcloud.

**Comment by SystemKeeper** [comment]
> Yeah, but this is out of scope for this issue here ;-) 

Yea, my life today 🙈 

> The signaling server should comply with whatever is returned by Nextcloud.

Yes, I agree with that in general. It just feels like a problem about to happen when we change that on the signaling side without a mitigation on the nextcloud side and we suddenly increase the number of requests significantly.

**Comment by fancycode** [comment]
Fixed by #783
